README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# SPLASH: Simple Process-Led Algorithms for Simulating Habitats <img src="https://raw.githubusercontent.com/villegar/splash/main/inst/images/logo.png" alt="logo" align="right" height=200px/>

<!-- badges: start -->
<!-- [![](https://img.shields.io/badge/devel%20version-1.0.2-yellow.svg)](https://github.com/villegar/splash) -->

[![](https://www.r-pkg.org/badges/version/splash?color=black)](https://cran.r-project.org/package=splash)
[![R build
status](https://github.com/villegar/splash/workflows/R-CMD-check/badge.svg)](https://github.com/villegar/splash/actions)
[![](https://img.shields.io/badge/doi-10.5281/zenodo.376293-blue.svg)](https://doi.org/10.5281/zenodo.376293)
<!-- badges: end -->

This program calculates bioclimatic indices and fluxes (radiation,
evapotranspiration, soil moisture) for use in studies of ecosystem
function, species distribution, and vegetation dynamics under changing
climate scenarios. Predictions are based on a minimum of required
inputs: latitude, precipitation, air temperature, and cloudiness.

## Installation

You can install the released version of splash from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("splash")
```

And the development version from [GitHub](https://github.com/) with:

<!-- You can install the development version from [GitHub](https://github.com/) with: -->

``` r
# install.packages("devtools")
devtools::install_github("villegar/splash")
```

## Reference

T. W. Davis, I. C. Prentice, B. D. Stocker, R. J. Whitley, H. Wang, B.
J. Evans, A. V. Gallego-Sala, M. T. Sykes, and W. Cramer, *Simple
process-led algorithms for simulating habitats (SPLASH): Robust indices
of radiation evapo-transpiration and plant-available moisture*,
Geoscientific Model Development, 2016. DOI:
<https://doi.org/10.5281/zenodo.376293>

Source: <https://bitbucket.org/labprentice/splash/>


NEWS.md
# splash 1.0.2
* Update the contact information of the package maintainer 

# splash 1.0.1
* Update link to website with documentation: 
https://villegar.github.io/splash -> https://splash.robertovillegas-diaz.com
* Added the `\doi` macro to the references for self-linking.

# splash 1.0.0

* Added a `NEWS.md` file to track changes to the package.
* Wrapped the standalone code into an R package.


R.ignore/test.R
# TEST 1: SOLAR ############################################################
solar <- calc_daily_solar(lat=37.7, n=172, elv=142, y=2000, sf=1, tc=23.0)
cat(sprintf("TEST 1---Solar values:\n"))
cat(sprintf("  kn: %d\n", solar$kN))
cat(sprintf("  nu: %0.6f degrees\n", solar$nu_deg))
cat(sprintf("  lambda: %0.6f degrees\n", solar$lambda_deg))
cat(sprintf("  rho: %0.6f\n", solar$rho))
cat(sprintf("  dr: %0.6f\n", solar$dr))
cat(sprintf("  delta: %0.6f degrees\n", solar$delta_deg))
cat(sprintf("  ru: %0.6f\n", solar$ru))
cat(sprintf("  rv: %0.6f\n", solar$rv))
cat(sprintf("  rw: %0.6f\n", solar$rw))
cat(sprintf("  hs: %0.6f degrees\n", solar$hs_deg))
cat(sprintf("  hn: %0.6f degrees\n", solar$hn_deg))
cat(sprintf("  tau_o: %0.6f\n", solar$tau_o))
cat(sprintf("  tau: %0.6f\n", solar$tau))
cat(sprintf("  Qn: %0.6f mol/m^2\n", solar$ppfd_mol.m2))
cat(sprintf("  Rnl: %0.6f w/m^2\n", solar$rnl_w.m2))
cat(sprintf("  Ho: %0.6f MJ/m^2\n", (1.0e-6) * solar$ra_j.m2))
cat(sprintf("  Hn: %0.6f MJ/m^2\n", (1.0e-6) * solar$rn_j.m2))
cat(sprintf("  Hnn: %0.6f MJ/m^2\n", (1.0e-6) * solar$rnn_j.m2))

# TEST 2: EVAP ############################################################
evap <- calc_daily_evap(lat=37.7, n=172, elv=142, y=2000, sf=1, tc=23.0, sw=0.9)
cat(sprintf("TEST 2---Evap values:\n"))
cat(sprintf("  s: %0.6f Pa/K\n", evap$s_pa.k))
cat(sprintf("  Lv: %0.6f MJ/kg\n", (1e-6) * evap$lv_j.kg))
cat(sprintf("  Patm: %0.6f bar\n", (1e-5) * evap$patm_pa))
cat(sprintf("  pw: %0.6f kg/m^3\n", evap$pw_kg.m3))
cat(sprintf("  gamma: %0.6f Pa/K\n", evap$gam_pa.k))
cat(sprintf("  Econ: %0.6f mm^3/J\n", (1e9) * evap$econ_m3.j))
cat(sprintf("  Cn: %0.6f mm\n", evap$cond_mm))
cat(sprintf("  rx: %0.6f\n", evap$rx))
cat(sprintf("  hi: %0.6f degrees\n", evap$hi_deg))
cat(sprintf("  EET: %0.6f mm\n", evap$eet_mm))
cat(sprintf("  PET: %0.6f mm\n", evap$pet_mm))
cat(sprintf("  AET: %0.6f mm\n", evap$aet_mm))

# TEST 3: SOIL MOISTURE (run one day) #########################################
t3 <- run_one_day(lat=37.7, elv=142, n=172, y=2000, wn=75, sf=1, tc=23, pn=5)
cat(sprintf("TEST 3---Soil moisture (run one day):\n"))
cat(sprintf("  Ho: %0.6f J/m2\n", t3$ho))
cat(sprintf("  Hn: %0.6f J/m2\n", t3$hn))
cat(sprintf("  PPFD: %0.6f mol/m2\n", t3$ppfd))
cat(sprintf("  EET: %0.6f mm/d\n", t3$eet))
cat(sprintf("  PET: %0.6f mm/d\n", t3$pet))
cat(sprintf("  AET: %0.6f mm/d\n", t3$aet))
cat(sprintf("  Cn: %0.6f mm/d\n", t3$cond))
cat(sprintf("  Wn: %0.6f mm\n", t3$wn))
cat(sprintf("  RO: %0.6f mm\n", t3$ro))

# # TEST 4:
daily_totals <- matrix(data=rep(0, 366), nrow=366, ncol=1)
daily_totals <- as.data.frame(daily_totals)
names(daily_totals) <- c("wn")
my_lat <- 37.7
my_elv <- 142
my_file <- system.file("extdata/example_data.csv", package = "splash")
my_data <- read_csv(my_file, 2000)
my_data$lat_deg <- my_lat
my_data$elv_m <- my_elv
daily_totals <- spin_up(my_data, daily_totals)
cat(sprintf("TEST 4---Spin-Up:\n"))
cat(sprintf("Day,Wn (mm)\n"))
for (i in seq(from=1, to=my_data$num_lines, by=1)) {
  cat(sprintf("%d,%0.6f\n", i, daily_totals$wn[i]))
}


R/data.R
# R version 3.2.3 (2015-12-10) -- "Wooden Christmas-Tree"
#
# data.R
#
# VERSION: 1.0
# LAST UPDATED: 2016-02-19
#
# ~~~~~~~~
# license:
# ~~~~~~~~
# Copyright (C) 2016 Prentice Lab
#
# This file is part of the SPLASH model.
#
# SPLASH is free software: you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 2.1 of the License, or
# (at your option) any later version.
#
# SPLASH is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with SPLASH.  If not, see <http://www.gnu.org/licenses/>.
#
# ~~~~~~~~~
# citation:
# ~~~~~~~~~
# T. W. Davis, I. C. Prentice, B. D. Stocker, R. J. Whitley, H. Wang, B. J.
# Evans, A. V. Gallego-Sala, M. T. Sykes, and W. Cramer, Simple process-led
# algorithms for simulating habitats (SPLASH): Robust indices of radiation
# evapo-transpiration and plant-available moisture, Geoscientific Model
# Development, 2016 (in progress)
#
# ~~~~~~~~~~~~
# description:
# ~~~~~~~~~~~~
# This script contains functions to handle the file IO for reading and writing
# data, i.e.:
#   read_csv(character fname, double y=-1)
#   read_txt(list my_data, character fname, character var, double y=-1)
#
# ~~~~~~~~~~
# changelog:
# ~~~~~~~~~~
# - created read_csv and read_txt functions [15.02.23]
#
#### DEFINE FUNCTIONS ########################################################
# ************************************************************************
# Name:     read_csv
# Inputs:   - character, file name (fname)
#           - double, year (y)
# Returns:  list object (data)
#             $file_name ............ file name
#             $sf ................... sunshine fraction
#             $tair ................. air temperature
#             $pn ................... precipitation
#             $num_lines ............ number of data points
#             $year ................. year of data
# Features: Reads all three daily input variables (sf, tair, and pn)
#           for a single year from a CSV file that includes a header
# ************************************************************************
#' Read CSV file
#'
#' Reads all three daily input variables (sf, tair, and pn)
#' for a single year from a CSV file that includes a header.
#'
#' @importFrom utils read.csv
#'
#' @param fname String, file name.
#' @param y Numeric, year.
#'
#' @return List with the following properties:
#' \describe{
#'   \item{$file_name}{File name.}
#'   \item{$sf}{Sunshine fraction.}
#'   \item{$tair}{Air temperature.}
#'   \item{$pn}{Precipitation.}
#'   \item{$num_lines}{Number of data points.}
#'   \item{$year}{Year of data.}
#' }
#' @export
read_csv <- function(fname, y = -1) {
  my_data <- list()
  my_data$file_name <- fname

  # Read data from CSV file and save to return list:
  DATA <- read.csv(fname)
  my_data$sf <- DATA$sf
  my_data$tair <- DATA$tair
  my_data$pn <- DATA$pn
  my_data$num_lines <- dim(DATA)[1]

  my_year <- y
  if (y == -1) {
    if (dim(DATA)[1] == 366) {
      my_year <- 2000
    } else if (dim(DATA)[1] == 365) {
      my_year <- 2001
    }
  }
  my_data$year <- my_year
  return (my_data)
}


# ************************************************************************
# Name:     read_txt
# Inputs:   - list object (my_data)
#           - character, file name (fname)
#           - character, variable name (var)
#           - double, year (y)
# Returns:  list object (my_data)
# Features: Reads plain text file (no header) of one of the input
#           arrays
# ************************************************************************
#' Read plain text file
#'
#' Reads plain text file (no header) of one of the input arrays.
#'
#' @param my_data List same as the output from \code{\link{read_csv}}.
#' @param fname String, file name.
#' @param var String, variable name.
#' @param y Numeric, year.
#'
#' @return List with the following properties:
#' \describe{
#'   \item{$file_name}{File name.}
#'   \item{$sf}{Sunshine fraction.}
#'   \item{$tair}{Air temperature.}
#'   \item{$pn}{Precipitation.}
#'   \item{$num_lines}{Number of data points.}
#'   \item{$year}{Year of data.}
#' }
#' @export
read_txt <- function(my_data, fname, var, y = -1) {
  my_data$file_name <- c(my_data$file_name, fname)
  DATA <- scan(fname)
  if (var == "sf") {
    my_data$sf <- DATA
  } else if (var == "tair") {
    my_data$tair <- DATA
  } else if (var == "pn") {
    my_data$pn <- DATA
  }
  my_data$num_lines <- length(DATA)

  my_year <- y
  if (y == -1) {
    if (length(DATA) == 366) {
      my_year <- 2000
    } else if (length(DATA) == 365) {
      my_year <- 2001
    }
  }
  my_data$year <- my_year
  return (my_data)
}


inst/const/const.R
# R version 3.2.3 (2015-12-10) -- "Wooden Christmas-Tree"
#
# const.R
#
# VERSION: 1.0-r1
# LAST UPDATED: 2016-05-27
#
# ~~~~~~~~
# license:
# ~~~~~~~~
# Copyright (C) 2016 Prentice Lab
#
# This file is part of the SPLASH model.
#
# SPLASH is free software: you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 2.1 of the License, or
# (at your option) any later version.
#
# SPLASH is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with SPLASH.  If not, see <http://www.gnu.org/licenses/>.
#
# ~~~~~~~~~
# citation:
# ~~~~~~~~~
# T. W. Davis, I. C. Prentice, B. D. Stocker, R. J. Whitley, H. Wang, B. J.
# Evans, A. V. Gallego-Sala, M. T. Sykes, and W. Cramer, Simple process-led
# algorithms for simulating habitats (SPLASH): Robust indices of radiation
# evapo-transpiration and plant-available moisture, Geoscientific Model
# Development, 2016 (in progress)
#
# ~~~~~~~~~~~~
# description:
# ~~~~~~~~~~~~
# This script contains the global constants defined in SPLASH.
#
# NOTE: orbital parameters: eccentricity, obliquity, and longitude of the
# perihelion, are assumed constant while they in fact vary slightly over time.
# There are methods for their calculation (e.g., Meeus, 1991). Eccentricity
# varies 0.005--0.072 and is decreasing at rate of 0.00004 per century.
# Obliquity varies 22.1--24.5 degrees with a period of ~41000 years.
#
# ~~~~~~~~~~
# changelog:
# ~~~~~~~~~~
# - updated values and references for ka and kR [14.10.31]
# - reduced list of constants [15.01.13]
# - updated kR and kTo values and references [15.03.24]
#
kA <- 107           # constant for Rl (Monteith & Unsworth, 1990)
kalb_sw <- 0.17     # shortwave albedo (Federer, 1968)
kalb_vis <- 0.03    # visible light albedo (Sellers, 1985)
kb <- 0.20          # constant for Rl (Linacre, 1968; Kramer, 1957)
kc <- 0.25          # constant for Rs (Linacre, 1968)
kCw <- 1.05         # supply constant, mm/hr (Federer, 1982)
kd <- 0.50          # constant for Rs (Linacre, 1968)
kfFEC <- 2.04       # from-flux-to-energy, umol/J (Meek et al., 1984)
kG <- 9.80665       # gravitational acceleration, m/s^2 (Allen, 1973)
kGsc <- 1360.8      # solar constant, W/m^2 (Kopp & Lean, 2011)
kL <- 0.0065        # adiabatic lapse rate, K/m (Cavcar, 2000)
kMa <- 0.028963     # molecular weight of dry air, kg/mol (Tsilingiris, 2008)
kMv <- 0.01802      # mol. weight of water vapor, kg/mol (Tsilingiris, 2008)
kSecInDay <- 86400  # number of seconds in a day
kPo <- 101325       # standard atmosphere, Pa (Allen, 1973)
kR <- 8.31447       # universal gas constant, J/mol/K (Moldover et al., 1988)
kTo <- 288.15       # base temperature, K (Berberan-Santos et al., 1997)
kWm <- 150          # soil moisture capacity, mm (Cramer-Prentice, 1988)
kw <- 0.26          # PET entrainment, (1+kw)*EET (Priestley-Taylor, 1972)
# pir <- pi / 180     # pi in radians

# # Paleoclimate variables:
# ke <- 0.01670       # eccentricity of earth's orbit, 2000CE (Berger 1978)
# keps <- 23.44       # obliquity of earth's elliptic, 2000CE (Berger 1978)
# komega <- 283       # lon. of perihelion, degrees, 2000CE (Berger, 1978)


R.ignore/hex-logo.R
hex_logo <- function(subplot = system.file("images/splash.png",
                                           package = "documentation"),
                     dpi = 600,
                     h_color = "#000000",
                     h_fill = "#FFFFFF",
                     output = system.file("images/logo.png",
                                          package = "splash"),
                     package = "SPLASH",
                     p_color = "#000000",
                     url = "https://github.com/villegar/splash",
                     u_size = 15) {
  hexSticker::sticker(subplot = subplot, package = package,
                      h_color = h_color,  h_fill = h_fill,
                      dpi = dpi,
                      s_x = 1.0, s_y = .9, s_width = .62, asp = .85,
                      p_x = 1.0, p_y = 1.58, p_size = 35, p_color = p_color,
                      url = url,
                      u_angle = 30, u_color = p_color, u_size = u_size,
                      filename = output)
}

hex_logo("inst/images/splash.png", output = "inst/images/logo.png", u_size = 9.3, dpi = 600)


R/splash.R
# R version 3.2.3 (2015-12-10) -- "Wooden Christmas-Tree"
#
# splash.R
#
# VERSION: 1.0
# LAST UPDATED: 2016-02-19
#
# ~~~~~~~~
# license:
# ~~~~~~~~
# Copyright (C) 2016 Prentice Lab
#
# This file is part of the SPLASH model.
#
# SPLASH is free software: you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 2.1 of the License, or
# (at your option) any later version.
#
# SPLASH is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with SPLASH.  If not, see <http://www.gnu.org/licenses/>.
#
# ~~~~~~~~~
# citation:
# ~~~~~~~~~
# T. W. Davis, I. C. Prentice, B. D. Stocker, R. J. Whitley, H. Wang, B. J.
# Evans, A. V. Gallego-Sala, M. T. Sykes, and W. Cramer, Simple process-led
# algorithms for simulating habitats (SPLASH): Robust indices of radiation
# evapo-transpiration and plant-available moisture, Geoscientific Model
# Development, 2016 (in progress)
#
# ~~~~~~~~~~~~
# description:
# ~~~~~~~~~~~~
# This script contains functions for running SPLASH for point-based data, i.e.:
#   spin_up(list mdat, list dtot)
#   quick_run(double lat, double elv, double n, double y, double wn, double sf,
#             double tc, double pn)
#   run_one_day(double lat, double elv, double n, double y, double wn,
#               double sf, double tc, double pn)
#
# ~~~~~~~~~~
# changelog:
# ~~~~~~~~~~
# - added fix to daily soil moisture when n and ny are 365 [15.01.27]
#
#### IMPORT SOURCES ##########################################################
# source("const.R")
# source("evap.R")


#### DEFINE FUNCTIONS ########################################################
# ************************************************************************
# Name:     spin_up
# Inputs:   - list, meteorological data (mdat)
#               $num_lines ..... double, length of meteorol. variable lists
#               $lat_deg ....... double latitude (degrees)
#               $elv_m ......... double, elevation (m)
#               $year .......... double, year
#               $sf ............ list, fraction of sunshine hours
#               $tair .......... list, mean daily air temperature (deg. C)
#               $pn ............ list, precipitation (mm/d)
#           - list, daily totals (dtot)
#               $wm ............ list, daily soil moisture (mm)
# Returns:  list, daily totals
# Features: Updates the soil moisture in daily totals until equilibrium
# Depends:  quick_run
# ************************************************************************
#' Calculate daily totals
#'
#' Calculate daily totals updating the soil moisture until equilibrium.
#'
#' @details The list with meteorological data, \code{mdat}, should have the
#' following components:
#' \itemize{
#'  \item num_lines ..... double, length of meteorol. variable lists
#'  \item lat_deg ....... double latitude (degrees)
#'  \item elv_m ......... double, elevation (m)
#'  \item year .......... double, year
#'  \item sf ............ list, fraction of sunshine hours
#'  \item tair .......... list, mean daily air temperature (deg. C)
#'  \item pn ............ list, precipitation (mm/d)
#' }
#'
#' The list with daily totals, \code{dtot}, should have the following component:
#' \itemize{
#'  \item wm ............ list, daily soil moisture (mm)
#' }
#'
#' @param mdat list with meteorological data (see the details section).
#' @param dtot list with daily totals (see the details section).
#'
#' @return list, daily totals
#' @export
#'
#' @examples
#' daily_totals <- matrix(data = rep(0, 366), nrow = 366, ncol = 1)
#' daily_totals <- as.data.frame(daily_totals)
#' names(daily_totals) <- c("wn")
#' my_file <- system.file("extdata/example_data.csv", package = "splash")
#' my_data <- splash::read_csv(my_file, 2000)
#' my_data$lat_deg <- 37.7
#' my_data$elv_m <- 142
#' daily_totals <- splash::spin_up(my_data, daily_totals)
#' cat(sprintf("Spin-Up:\n"))
#' for (i in seq(from = 1, to = my_data$num_lines, by = 1)) {
#'   if (i == 1) cat(sprintf("Day\tWn (mm)\n"))
#'   cat(sprintf("%d\t%0.6f\n", i, daily_totals$wn[i]))
#' }
spin_up <- function(mdat, dtot) {
  # Run one year:
  for (i in seq(from = 1, to = mdat$num_lines, by = 1)) {
    if (i == 1) {
      wn <- dtot$wn[mdat$num_lines]
    } else {
      wn <- dtot$wn[i - 1]
    }
    my_vals <- quick_run(mdat$lat_deg,
                         mdat$elv_m,
                         i,
                         mdat$year,
                         wn,
                         mdat$sf[i],
                         mdat$tair[i],
                         mdat$pn[i])
    dtot$wn[i] <- my_vals$sm
  }

  # Calculate the change:
  start_sm <- dtot$wn[1]
  end_vals <- quick_run(mdat$lat_deg, mdat$elv_m, 1, mdat$year,
                        dtot$wn[mdat$num_lines], mdat$sf[1], mdat$tair[1],
                        mdat$pn[1])
  diff_sm <- abs(end_vals$sm - start_sm)

  # Equilibrate:
  spin_count <- 1
  while (diff_sm > 1.0) {
    for (i in seq(from = 1, to = mdat$num_lines, by = 1)) {
      if (i == 1) {
        wn <- dtot$wn[mdat$num_lines]
      } else {
        wn <- dtot$wn[i - 1]
      }
      my_vals <- quick_run(mdat$lat_deg, mdat$elv_m, i, mdat$year, wn,
                           mdat$sf[i], mdat$tair[i], mdat$pn[i])
      dtot$wn[i] <- my_vals$sm
    }

    # Calculate the change:
    start_sm <- dtot$wn[1]
    end_vals <- quick_run(mdat$lat_deg, mdat$elv_m, 1, mdat$year,
                          dtot$wn[mdat$num_lines], mdat$sf[1],
                          mdat$tair[1], mdat$pn[1])
    diff_sm <- abs(end_vals$sm - start_sm)
    spin_count <- spin_count + 1
  }
  cat(paste("Spun", spin_count, "years\n"))
  return(dtot)
}


# ************************************************************************
# Name:     quick_run
# Inputs:   - double, latitude, deg (lat)
#           - double, elevation, m (elv)
#           - double, day of year (n)
#           - double, year (y)
#           - double, daily soil moisture content, mm (wn)
#           - double, daily fraction of bright sunshine (sf)
#           - double, daily air temperature, deg C (tc)
#           - double, daily precipitation, mm (pn)
# Returns:  list
#             $sm - soil moisture, mm
#             $ro - runoff, mm
# Features: Returns daily soil moisture and runoff
# Depends:  evap
# ************************************************************************
#' Calculate daily soil moisture and runoff
#'
#' @inheritParams calc_daily_evap
#' @param pn double, daily precipitation, mm/day.
#' @param kCw double, supply constant, mm/hr.
#'     Default: \eqn{1.05} (Federer, 1982)
#' @param kWm double, soil moisture capacity, mm.
#'     Default: \eqn{150} (Cramer-Prentice, 1988)
#'
#' @return Returns daily soil moisture and runoff.
#' @keywords internal
#'
#' @references
#' Cramer, W. and Prentice, I.C., 1988. Simulation of regional soil moisture
#' deficits on a European scale. Norsk Geografisk Tidsskrift - Norwegian Journal
#' of Geography, 42(2-3), pp.149–151. \doi{10.1080/00291958808552193}
#'
#' Federer, C.A., 1982. Transpirational supply and demand: plant, soil, and
#' atmospheric effects evaluated by simulation. Water Resources Research, 18(2),
#' pp.355-362. \doi{10.1029/WR018i002p00355}
quick_run <- function(lat, elv, n, y, wn, sf, tc, pn, kCw = 1.05, kWm = 150) {
  # Calculate evaporative supply (mm/hr)
  sw <- kCw * wn / kWm

  # Compute daily radiation and evaporations values:
  ET <- calc_daily_evap(lat, n, elv, y, sf, tc, sw)

  # Update daily soil moisture:
  sm <- wn + pn + ET$cond_mm - ET$aet_mm

  if (sm > kWm) {
    # Bucket is full:
    # - set soil moisture to capacity
    # - add remaining water to runoff
    ro <- sm - kWm
    sm <- kWm
  } else if (sm < 0) {
    # Bucket is empty:
    # - set runoff and soil moisture equal to zero
    ro <- 0
    sm <- 0
  } else {
    ro <- 0
  }

  rval <- list()
  rval$sm <- sm
  rval$ro <- ro
  return(rval)
}


# ************************************************************************
# Name:     run_one_day
# Inputs:   - double, latitude, deg (lat)
#           - double, elevation, m (elv)
#           - double, day of year (n)
#           - double, year (y)
#           - double, daily soil moisture content, mm (wn)
#           - double, daily fraction of bright sunshine (sf)
#           - double, daily air temperature, deg C (tc)
#           - double, daily precipitation, mm (pn)
# Returns:  list
#             $ho - daily solar irradiation, J/m2
#             $hn - daily net radiation, J/m2
#             $ppfd - daily PPFD, mol/m2
#             $cond - daily condensation water, mm
#             $eet - daily equilibrium ET, mm
#             $pet - daily potential ET, mm
#             $aet - daily actual ET, mm
#             $wn - daily soil moisture, mm
#             $ro - daily runoff, mm
# Features: Runs SPLASH at a single location for one day.
# Depends:  evap
# ************************************************************************
#' Runs SPLASH at a single location for one day
#'
#' @param wn double, daily soil moisture content, mm (wn).
#' @inheritParams quick_run
#'
#' @return List with the following components:
#' \itemize{
#'  \item ho .......... daily solar irradiation, J/m2
#'  \item hn .......... daily net radiation, J/m2
#'  \item ppfd ........ daily PPFD, mol/m2
#'  \item cond ........ daily condensation water, mm
#'  \item eet ......... daily equilibrium ET, mm
#'  \item pet ......... daily potential ET, mm
#'  \item aet ......... daily actual ET, mm
#'  \item wn .......... daily soil moisture, mm
#'  \item ro .......... daily runoff, mm
#' }
#'
#' @export
#'
#' @references
#' Cramer, W. and Prentice, I.C., 1988. Simulation of regional soil moisture
#' deficits on a European scale. Norsk Geografisk Tidsskrift - Norwegian Journal
#' of Geography, 42(2-3), pp.149–151. \doi{10.1080/00291958808552193}
#'
#' Federer, C.A., 1982. Transpirational supply and demand: plant, soil, and
#' atmospheric effects evaluated by simulation. Water Resources Research, 18(2),
#' pp.355-362. \doi{10.1029/WR018i002p00355}
#'
#' @examples
#' soil <- run_one_day(lat = 37.7,
#'                     elv = 142,
#'                     n = 172,
#'                     y = 2000,
#'                     wn = 75,
#'                     sf = 1,
#'                     tc = 23,
#'                     pn = 5)
#' cat(sprintf("Soil moisture (run one day):\n"))
#' cat(sprintf("  Ho: %0.6f J/m2\n", soil$ho))
#' cat(sprintf("  Hn: %0.6f J/m2\n", soil$hn))
#' cat(sprintf("  PPFD: %0.6f mol/m2\n", soil$ppfd))
#' cat(sprintf("  EET: %0.6f mm/d\n", soil$eet))
#' cat(sprintf("  PET: %0.6f mm/d\n", soil$pet))
#' cat(sprintf("  AET: %0.6f mm/d\n", soil$aet))
#' cat(sprintf("  Cn: %0.6f mm/d\n", soil$cond))
#' cat(sprintf("  Wn: %0.6f mm\n", soil$wn))
#' cat(sprintf("  RO: %0.6f mm\n", soil$ro))
run_one_day <- function(lat, elv, n, y, wn, sf, tc, pn, kCw = 1.05, kWm = 150) {
  # Return values
  rvals <- list()

  # Calculate evaporative supply (mm/hr)
  sw <- kCw*wn/kWm

  # Compute daily radiation and evaporations values:
  ET <- calc_daily_evap(lat, n, elv, y, sf, tc, sw)
  rvals$ho <- ET$ra_j.m2
  rvals$hn <- ET$rn_j.m2
  rvals$ppfd <- ET$ppfd_mol.m2
  rvals$cond <- ET$cond_mm
  rvals$eet <- ET$eet_mm
  rvals$pet <- ET$pet_mm
  rvals$aet <- ET$aet_mm

  # Update daily soil moisture:
  sm <- wn + pn + ET$cond_mm - ET$aet_mm

  print( paste( "in run_one_day: pn =", pn))

  if (sm > kWm) {
    # Bucket is full:
    # - set soil moisture to capacity
    # - add remaining water to runoff
    ro <- sm - kWm
    sm <- kWm
  } else if (sm < 0) {
    # Bucket is empty:
    # - reduce actual ET by discrepancy amount
    # - set runoff and soil moisture equal to zero
    rvals$aet <- rvals$aet + sm
    ro <- 0
    sm <- 0
  } else {
    ro <- 0
  }

  rvals$wn <- sm
  rvals$ro <- ro
  return(rvals)
}
